contracts/p-plus-epsilon.sol

```diff
@@ -0,0 +1,145 @@
+  }
+
+  /** @dev The attacker can withdraw their PNK from here after the bribe has been settled.
+   */
+  function withdrawAttacker(){
+    require(settled);
+
+    if (balance > remainingWithdraw) {
+      // The attacker can withdraw the remaning balance of PNK after the contract has been settled.
+      require(pinakion.transfer(attacker, balance - remainingWithdraw));
+    }
+  }
+
+  /** @dev Settles the p + e bribe with the jurors.
+   * If the dispute is ruled differently from desiredOutcome:
+   *    The jurors who voted for desiredOutcome receive p + d + e in rewards from this contract.
+   * If the dispute is ruled as in desiredOutcome:
+   *    The jurors don't receive anything from this contract.
+   */
+  function settle() public {
+    require(court.disputeStatus(disputeID) ==  Arbitrator.DisputeStatus.Solved);
+    require(!settled); // This function can be executed only once.
+
+    settled = true; // settle the bribe
+
+    // From the dispute we get the # of appeals and available choices
+    var (, , appeals, choices, , , ,) = court.disputes(disputeID);
+
+    if (court.currentRuling(disputeID) != desiredOutcome){
+      // Calculate the redistribution amounts.
+      uint amountShift = court.getStakePerDraw();
+      uint winningChoice = court.getWinningChoice(disputeID, appeals);
+
+      // Rewards are calculated as per the one shot token reparation.
+      for (uint i=0; i < appeals; i++){ // Loop each appeal and each vote.
+
+        // Note that we don't check if the result was a tie becuse we are getting a funny compiler error: "stack is too deep" if we check.
+        // TODO: Account for ties
+        if (winningChoice != 0){
+          // votesLen is the lenght of the votes per each appeal. There is no getter function for that, so we have to calculate it here.
+          // We must end up with the exact same value as if we would have called dispute.votes[i].length
+          uint votesLen = 0;
+          for (uint c = 0; c < choices; c++) { // Iterate for each choice of the dispute.
+            votesLen += court.getVoteCount(disputeID, i, c);
+          }
+
+          uint totalToRedistribute = 0;
+          uint nbCoherent = 0;
+
+          // Now we will use votesLen as a substitute for dispute.votes[i].length
+          for (uint j=0; j < votesLen; j++){
+            uint voteRuling = court.getVoteRuling(disputeID, i, j);
+            address voteAccount = court.getVoteAccount(disputeID, i, j);
+
+            if (voteRuling != winningChoice){
+              totalToRedistribute += amountShift;
+
+              if (voteRuling == desiredOutcome){ // If the juror voted as we desired.
+                // Transfer this juror back the penalty.
+                withdraw[voteAccount] += amountShift;
+                remainingWithdraw += amountShift;
+              }
+            } else {
+              nbCoherent++;
+            }
+          }
+          if (nbCoherent > 0){
```

Comment:
I don't understand what you are doing there.
You need to give (amountShift+epsilon) * amountDraw to everyone who voted for desiredOutcome if desiredOutcome is not winning.

Reply:
Ok, in addition you add what they would have won from incoherent voters would they have been. So this seems correct.